src/webui/webapplication.cpp

```diff
@@ -173,45 +98,123 @@ namespace
+                translation.replace("'", "&#39;");
+                translation.replace("\"", "&#34;");
+
+                data.replace(i, regex.matchedLength(), translation);
+                i += translation.length();
+            }
+            else {
+                found = false; // no more translatable strings
+            }
+        }
+    }
 }
 
-void WebApplication::action_public_index()
+WebApplication::WebApplication(QObject *parent)
+    : AbstractWebApplication(parent)
+    , m_isAltUIUsed(false)
 {
-    QString path;
-
-    if (!args_.isEmpty()) {
-        if (args_.back() == "favicon.ico")
-            path = ":/icons/skin/qbittorrent16.png";
-        else
-            path = WWW_FOLDER + args_.join("/");
+    for (int i = metaObject()->methodOffset(); i < metaObject()->methodCount(); ++i) {
+        QMetaMethod method = metaObject()->method(i);
+        if (QLatin1String(method.tag()) == QLatin1String("WEBAPI_PUBLIC"))
+            m_publicAPIs << method.name();
+        else if (QLatin1String(method.tag()) == QLatin1String("WEBAPI_PRIVATE"))
+            m_privateAPIs << method.name();
     }
 
-    printFile(path);
+    configure();
+    connect(Preferences::instance(), &Preferences::changed, this, &WebApplication::configure);
 }
 
-void WebApplication::action_public_webui()
+void WebApplication::sendWebUIFile()
 {
-    if (!sessionActive())
-        printFile(PRIVATE_FOLDER + "login.html");
-    else
-        printFile(PRIVATE_FOLDER + "index.html");
+    const QStringList pathItems {request().path.split('/', QString::SkipEmptyParts)};
+    if (pathItems.contains(".") || pathItems.contains("..")) {
+        status(404, "Not Found");
+        return;
+    }
+
+    if ((!m_isAltUIUsed) && (request().path.startsWith(PATH_PREFIX_IMAGES))) {
+        printFile(QLatin1String(":/icons/") + request().path.mid(PATH_PREFIX_IMAGES.size()));
+        header(Http::HEADER_CACHE_CONTROL, MAX_AGE_MONTH);
+        return;
+    }
+
+    const QString path {
+        (request().path != QLatin1String("/")
+                ? request().path
+                : (sessionActive()
+                   ? QLatin1String("/index.html")
+                   : QLatin1String("/login.html")))
+    };
+
+    QString localPath {
+        m_rootFolder
+                + (sessionActive() ? PRIVATE_FOLDER : PUBLIC_FOLDER)
+                + path
+    };
+
+    QFileInfo fileInfo {localPath};
+
+    if (!fileInfo.exists() && sessionActive()) {
+        // try to send public file if there is no private one
+        localPath = m_rootFolder + PUBLIC_FOLDER + path;
+        fileInfo.setFile(localPath);
+    }
+
+    while (fileInfo.filePath() != m_rootFolder) {
+        if (fileInfo.isSymLink()) {
+            status(500, "Internal Server Error");
+            print(tr("Symlinks inside WWW root folder are forbidden."), Http::CONTENT_TYPE_TXT);
+            return;
+        }
+
+        fileInfo.setFile(fileInfo.path());
```

Comment:
Can't understand what does this line do.

Reply:
Reset QFileInfo object to another path (parent path of current file).